### PR TITLE
Add failing test fixture for AddArrayReturnDocTypeRector

### PR DESCRIPTION
**Pull request by staabm**
# Failing Test for AddArrayReturnDocTypeRector

Based on https://getrector.org/demo/1ec63ff9-ccd7-654c-9efd-5f80c7c1a629

### CODE REVIEW

File: rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/base_find.php.inc [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    public function find()
+    {
+    }    
+}
+
+class Finder extends BaseFind
+{
+    /**
+     * @return static|false
+     */
```
Comment (staabm) [inline rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/base_find.php.inc]: atm rector is infering a different type here and therefore looses precision

ACTUAL:
```diff
+    /**
+     * @return bool|$this
+     */
```

Comment (samsonasik) [reply rules-tests/TypeDeclaration/Rector/ClassMethod/AddArrayReturnDocTypeRector/Fixture/base_find.php.inc]: probably related with https://github.com/rectorphp/rector-src/pull/1521

**Comment by samsonasik** [comment]
@staabm it possibly already resolved in latest main, could you rebase? Thank you.

**Comment by samsonasik** [comment]
The issue seems still exists https://getrector.org/demo/1ec6b2ec-da62-6d86-92ef-2be0d799690c

**Comment by staabm** [comment]
rebased.

**Comment by TomasVotruba** [comment]
I'm looking into this. In the case of `false` to `bool` there is nothing we can do. Let me explain why.

Type resolver works with exact provided values of specific methods.

```diff
public function isEnabled()
{
    return false;
}
```

After your change, this would be interpreted as `false` type, but is clearly a `bool` type.


<br>

What can you improve on your side?

* use strict type declaration  - `sometMethod(): false|string`
* use nullable return - `someMethod(): ?string` - static tools and PHP generaly understand this better, as nullable is with us sine PHP 7.1, while `false` since PHP 8.0
* the `false` is `boolean` type, so "something" is `string` type (though static in this case, but the context depends on user), so is more vague to read

If that doesn't fit your taste, skip the rule + file here.
---

I'm looking into the `static` type

**Comment by TomasVotruba** [comment]
The `static` to `ObjectType` retype comes directly from PHPStan scope:

image

As we rely on the type system, it would have to be changed from `ObjectType` to `StaticType` in PHPStan itself.

**Comment by TomasVotruba** [comment]
On another topic... I've noticed this rule is called "AddArray...". It should not handle these union types at all :) I've probably used incorrect if condition while refactoring from our types to PHPStan types, back in the days.

This PR fixes it https://github.com/rectorphp/rector-src/pull/1639

**Comment by staabm** [comment]
thanks!

> As we rely on the type system, it would have to be changed from `ObjectType` to `StaticType` in PHPStan itself.

@TomasVotruba  do you think this is a bug in phpstan worth reporting?

**Comment by TomasVotruba** [comment]
It's a behavior that makes sense in PHPStan scope. As the type is all that matters.
But for copying type the source can be various source. It's worth trying

**Comment by staabm** [comment]
I see. since you have all the details at hand, could you open a issue?

**Comment by TomasVotruba** [comment]
It's over my capacity, I'm fully focused here

**Comment by staabm** [comment]
> Type resolver works with exact provided values of specific methods.
> 
> ```diff
> public function isEnabled()
> {
>     return false;
> }
> ```
> 
> After your change, this would be interpreted as `false` type, but is clearly a `bool` type.


IMO this is not clearly a `BooleanType`, but a `ConstantBooleanType(false)` and `false` is a known type to php-src, it seems
https://wiki.php.net/rfc/union_types_v2#false_pseudo-type

**Pull request closed without merging**